mypy/checker.py

```diff
@@ -3399,6 +3399,14 @@ def check_multi_assignment(
             if len(relevant_items) == 1:
                 rvalue_type = get_proper_type(relevant_items[0])
 
+        if isinstance(rvalue_type, FunctionLike) and rvalue_type.is_type_obj():
+            ret_type = get_proper_type(rvalue_type.items[0].ret_type)
```

Comment:
This code does not support `@overload` on `__iter__`.

For example:

```python
from typing import Iterator, Type, TypeVar, Generic, overload

T = TypeVar('T')

class Meta(type, Generic[T]):
    @overload  # special case
    def __iter__(cls: Meta[int]) -> Iterator[None]: ...
    @overload  # special case
    def __iter__(cls: Meta[T]) -> Iterator[T]: ...
```

Do we care to support this very rare case?

Reply:
It looks like there are also issues for overloaded `__iter__` methods outside of the metaclass case I'm trying to tackle in this PR:

https://mypy-play.net/?mypy=latest&python=3.11&gist=fea665e2d519eeb17fc9af8ed8ee6086

Okay if I open a new issue for this and say it's out of scope here? :)

Reply:
I am ok with it :)

Reply:
I've opened this issue for the more general problem with overloaded `__iter__` methods:

- #14811

An overloaded `__iter__` method in a metaclass wouldn't be supported anyway; mypy hates generic metaclasses:

```python
from typing import AnyStr, Generic, overload, Iterator

class Meta(type, Generic[AnyStr]):
    x: AnyStr
    @overload
    def __iter__(self: Meta[str]) -> Iterator[bytes]: ...
    @overload
    def __iter__(self: Meta[bytes]) -> Iterator[str]: ...
    def __iter__(self) -> Iterator[str | bytes]:
        if isinstance(self.x, str):
            yield from [b"a", b"b", b"c"]
        yield from ["a", "b", "c"]

class Foo(metaclass=Meta[str]): ...  # error: Dynamic metaclass not supported for "Foo"  [misc]

a, b, c = Foo  # error: "Type[Foo]" object is not iterable  [misc]
reveal_type(a)  # error: Cannot determine type of "a"  [has-type]  # note: Revealed type is "Any"
```

Reply:
> Okay if you leave a comment in code!

Comment added in 73f92eb